File: Brainfuck/brainfuck.py
#!/usr/bin/env python3
# 
# Usage: ./brainfuck.py file

import sys

class _Getch:
    """Gets a single character from standard input. Does not echo to the
screen."""
    def __init__(self):
        try:
            self.impl = _GetchWindows()
        except ImportError:
            self.impl = _GetchUnix()

    def __call__(self):
        return self.impl()

class _GetchUnix:
    def __init__(self):
        import tty

    def __call__(self):
        import sys, tty, termios
        fd = sys.stdin.fileno()
        old_settings = termios.tcgetattr(fd)
        try:
            tty.setraw(sys.stdin.fileno())
            ch = sys.stdin.read(1)
        finally:
            termios.tcsetattr(fd, termios.TCSADRAIN, old_settings)
        return ch

class _GetchWindows:
    def __init__(self):
        import msvcrt

    def __call__(self):
        import msvcrt
        return msvcrt.getch()



def execute(filename):
    with open(filename, 'r') as f:
        evaluate(f.read())


def evaluate(code):
    code = cleanup(code)
    bracemap = buildbracemap(code)

    cells, codeptr, cellptr = [0], 0, 0

    while codeptr < len(code):
        command = code[codeptr]
        if command == ">":
            cellptr += 1
            if cellptr == len(cells): cells.append(0)
        elif command == "<":
            cellptr = 0 if cellptr <= 0 else cellptr - 1
        elif command == "+":
            cells[cellptr] = cells[cellptr] + 1 if cells[cellptr] < 255 else 0
        elif command == "-":
            cells[cellptr] = cells[cellptr] - 1 if cells[cellptr] > 0 else 255
        elif command == "[" and cells[cellptr] == 0:
            codeptr = bracemap[codeptr]
        elif command == "]" and cells[cellptr] != 0:
            codeptr = bracemap[codeptr]
        elif command == ".":
            sys.stdout.write(chr(cells[cellptr]))
        elif command == ",":
            getch = _Getch()
            cells[cellptr] = ord(getch())
        
        codeptr += 1


def cleanup(code):
    out = filter(lambda x: x in ['>', '<', '+', '-', '[', ']', '.', ','], code)
    return list(out)


def buildbracemap(code):
    temp_bracestack, bracemap = [], {}
    for pos, cmd in enumerate(code):
        if cmd == "[": temp_bracestack.append(pos)
        if cmd == "]":
            start = temp_bracestack.pop()
            bracemap[start] = pos
            bracemap[pos] = start

    return bracemap


def main():
    if len(sys.argv) == 2:
        execute(sys.argv[1])
    else:
        print("Usage:", sys.argv[0], "filename")


if __name__ == "__main__":
    main()

File: Brainfuck/brainfuck.cpp
#include <stdlib.h>
#include <iostream>
#include <string>
#include <fstream>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

#define usage "usage: brainfuck infile"


map<int, int> get_bracemap(vector<char> code) {
    vector<int> temp_bm;
    map<int, int> bm;
    for (int i=0; i<code.size(); i++) {
        if (code[i] == '[') {
            temp_bm.push_back(i);
        } else if (code[i] == ']') {
            int start = temp_bm.back();
            temp_bm.pop_back();
            bm[start] = i;
            bm[i] = start;
        }
    }

    return bm;
}


void evaluate(vector<char> code, map<int, int> bm) {
    vector<int> cells;
    cells.push_back(0);

    int codeptr = 0;
    int cellptr = 0;

    char cmd;
    while (codeptr < code.size()) {
        cmd = code[codeptr];
        switch (cmd) {
            case '>':
                if (++cellptr == cells.size()) cells.push_back(0);
                break;
            case '<':
                cellptr = cellptr <= 0 ? 0 : --cellptr;
                break;
            case '+':
                cells[cellptr] = cells[cellptr] < 255 ? ++cells[cellptr] : 0;
                break;
            case '-':
                cells[cellptr] = cells[cellptr] > 0 ? --cells[cellptr] : 255;
                break;
            case '[':
                if (cells[cellptr] == 0) codeptr = bm[codeptr];
                break;
            case ']':
                if (cells[cellptr] != 0) codeptr = bm[codeptr];
                break;
            case '.':
                cout << (char)cells[cellptr];
                break;
            case ',':
                char c;
                cin >> c;
                cells[cellptr] = (int)c;
                break;
        }

        codeptr++;
    }
}


void execute(char* fn) {
    if (!fn) {
        cerr << usage << endl;
        exit(1);
    }

    vector<char> code;
    
    ifstream file(fn);
    if (file.is_open()) {
        char bf_chars_arr[] = {'<', '>', '+', '-', '[', ']', '.', ','};
        vector<char> bf_chars(bf_chars_arr, bf_chars_arr + sizeof(bf_chars_arr) / sizeof(bf_chars_arr[0]));
        
        string line;
        while (getline(file, line)) {
            for (char c : line) {
                if (find(bf_chars.begin(), bf_chars.end(), c) != bf_chars.end()) {
                    code.push_back(c);
                }
            }
        }

        file.close();

        evaluate(code, get_bracemap(code));
    } else {
        cerr << "Unable to open file " << fn << endl;
        exit(1);
    }
}


int main(int argc, char* argv[]) {
    if (argc == 2) {
        execute(argv[1]);
        return 0;
    } else {
        cout << usage << endl;
        return 1;
    }
}
